만국박람회 [STEP 2] willy.kim, hayden.chae

cms7371 opened the pull request:
# 구현 내용
### ExpoViewController
- ScrollView와 StackView를 이용하여 메인 화면을 구성했습니다
- 데이터는 JasonParser를 이용하여 모델을 불러와 View를 업데이트 하였습니다
### HeritageTableViewController
- UITableViewController, CustomTableViewCell을 이용하여 출품작 화면을 구성했습니다
- JasonParser를 이용하여 불러온 아이템 데이터를 이용하여 TableCell을 구현
- 셀 터치동작시 다음 화면으로 모델을 넘겨주는 방식으로 플로우를 만들었습니다
### HerigtageViewController
- 넘겨받은 데이터를 이용하여 출품작 상세정보를 표현하는 화면을 구현하였습니다
### JasonParser
- enum을 이용하여 로드될 파일을 정의하고, 이를 모델로 파싱해주는 클래스를 만들었습니다
- JasonDecoder가 반복적으로 이용될 것을 생각하여, static util class가 아닌 싱글톤으로 동작하도록 했습니다.
# 궁금한 점
- 출품작 상세정보 표시에서 한가지 문제점을 발견하였습니다.
다른 이미지는 문제없이 instrinsic size를 통해 스택 뷰 안에 적절한 사이즈로 안착되었지만
해금과 나전칠기가 위아래 마진이 커지는 현상이 발생했습니다.
열심히 분석해본 결과 해금과 나전칠기 해상도가 높아 아주 큰 instrinsic size를 가졌고,
이것이 stack view 내의 제약과 상충되어 width만이 줄어들어 발생하는 문제임을 알게 되었습니다.
stackview에서 intrinsic size 요구 -> 큰 image size 반환 -> 좌우 한계로 width만 줄여서 layoutSubview
의 프로세스로 진행된다는 것을 추측했고 
storyboard 상에서는 이를 해결할 수 없다는 것이 분명하여 코드로 image의 비율을 계산해
ratio constrait로 추가해주는 방법을 이용하게 되었습니다.
야곰님께서는 다른 제약을 설정하지 말라하셨기에 저희가 선택한 방법 외에 다른 솔루션이 있는지 궁금합니다.
<img src="https://user-images.githubusercontent.com/65817235/149099109-6be65110-c825-4cb8-9e7f-29e413e7981c.png"   height="400"/><img src="https://user-images.githubusercontent.com/65817235/149098436-01f83fc2-b7a0-47cc-9141-a994d710abc2.png"  width="200" height="400"/><img src="https://user-images.githubusercontent.com/65817235/149098503-ea4dbdab-edd5-410a-819a-cd1dce56acee.png"  width="200" height="400"/>
```swift
lazy var heritages: [Heritage] = {
        do {
            return try JsonParser.shared.fetchData(from: .expo1900Items)
        } catch {
            fatalError("Failed to decode json data")
        }
    }()
```
- 위의 코드는 HeritageTableViewController.swift에서의 출품작 모델 초기화 코드입니다. 처음에 ViewDidLoad에서 진행을 하려다가 토의한 결과 위와 같은 방법으로 초기화를 진행하였습니다.
위와 같은 방법은 어느 시점에서 초기화가 일어나는지 확실치 않다는 점을 제외하고는 문제가 없어보이는데 이와 같은 방식이 괜찮은지 궁금합니다.





cc. @zdodev

cms7371 committed:
feat: Step2 메인 화면 구성

- Storyboard에서 Scroll View, Stack View를 이용한 메인 화면 구성
- contstraint가 꼬여 스크롤이 되지 않는 문제 발생

wilybear committed:
feat: STEP2 메인 화면 데이터 표시 구현

- NavigationViewController를 통한 앱 플로우 구성
- ExpoViewController 메인 엑스포 정보 표시
- DataManager: 에셋의 JSON 파일에서 모델로 파싱하여 전달하는 싱글톤 객체
- Int+Extension: 숫자의 코마를 추가할 수 있도록 확장

cms7371 committed:
feat: Step2 출품작 TableView 구현

- Storyboard에서 TableViewController를 통해 출품작 화면 구현
- DataManage에 출품작 JSON 파싱 메서드 getMasterpieces 구현
- MasterpieceTableViewController로 불러온 데이터 표시 기능 구현
- MasterpieceTableViewCell에서 데이터 표시 커스텀 셀 정의

wilybear committed:
feat: step2 출품작 상세화면 구현

- HeritageViewController: 출품작 상세화면 구현
- HeritageTableViewController: 셀 선택 동작 구현
- DataManager -> JsonParser로 이름 변경
- Masterpiece -> Heritage로 이름 변경
- 파일 디렉토리 정리

wilybear committed:
refactor: step2 출품작 이미지 빈 공간 제거, JsonParser 리펙터

- 출품작 이미지 위 아래로 과도한 빈공간이 생기는 현상 수정
- JsonParser의 중복 코드 제너릭으로 수정, enum으로 jsonfile 정리

zdodev commented on Expo1900/Expo1900/Model/Heritage.swift:
```diff
@@ -1,13 +1,13 @@
 import Foundation
 
-struct Masterpiece: Codable {
+struct Heritage: Codable {
```
현재 스텝에서는 해당 모델에 대해서 `Encoding`하는 부분이 보이지 않습니다.
`Codable`이 어떤 프로토콜이 합성된 타입인지 알아보면 좋을 것 같습니다.😁
`Decoding`만 한다면 `Encoding`에 대한 책임을 없애보는 것은 어떨까요?🤔

cms7371 replied:
처음에 생각없이 Codable로 했었는데 좋은 제안인 것 같습니다! 감사합니다!!

zdodev commented on Expo1900/Expo1900/Utils/Int+Extension.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+extension Int {
+    func withCommas() -> String {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        return numberFormatter.string(from: NSNumber(value: self)) ?? "0"
+    }
+}
```
comma 포맷 extension을 정의해보셨네요. 좋습니다.👍

zdodev commented on Expo1900/Expo1900/Utils/JsonParser.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    private init() { }
+    
+    func fetchData<T: Codable>(from fileName: JsonFileList) throws -> T{
```
'{' 괄호를 붙여서 사용하는 구문도 있고 띄어쓴 구문도 있습니다. 하나의 통일된 방식으로 사용해보는 것은 어떨까요?🤔

cms7371 replied:
공백 몇 부분 깜빡한 것 같습니다. 일관되게 습관을 들여보도록 하겠습니다!

zdodev commented on Expo1900/Expo1900/Utils/JsonParser.swift:
```diff
@@ -0,0 +1,29 @@
+    private init() { }
+    
+    func fetchData<T: Codable>(from fileName: JsonFileList) throws -> T{
+        guard let jsonData = NSDataAsset(name: fileName.rawValue)?.data else { fatalError("Json file doesn't exist") }
+        let data = try jsonDecoder.decode(T.self, from: jsonData)
```
`NSDataAsset`를 통해 가져온 데이터가 아직은 json 포맷이 아니라는 생각인데 `jsonData` 대신 다른 이름을 사용해보는 것은 어떨까요?
아래의 `data` 이름과 바꿔보면 더 역할에 적합한 이름이지 않을까 싶었습니다. 의견이 궁금합니다~😁

cms7371 replied:
저희도 다시 읽어보니 내부 변수명이 적절하지 않다고 생각되어 리팩터 진행했습니다

zdodev commented on Expo1900/Expo1900/View/HeritageTableViewCell.swift:
```diff
@@ -0,0 +1,24 @@
+    @IBOutlet weak var heritageTitleLabel: UILabel!
+    @IBOutlet weak var heritageShortDescriptionLabel: UILabel!
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+    }
```
아래 코드는 특별히 사용하지 않는 것으로 보이는데, 사용하지 않는 코드는 제거하는 것이 어떨까요?🤔

zdodev commented on Expo1900/Expo1900/ViewController/ExpoViewController.swift:
```diff
@@ -0,0 +1,50 @@
+    @IBOutlet weak var durationLabel: UILabel!
+    @IBOutlet weak var descriptionTextView: UITextView!
+    
+    lazy var expoModel: InternationalExposition = {
```
`lazy` 키워드를 잘 활용해보셨네요. 좋습니다.👍
`lazy` 사용해서 실제로 사용할 때 메모리에 할당한다는 측면에 현재로서는 좋은 방식이라고 생각이 듭니다~

zdodev commented on Expo1900/Expo1900/ViewController/ExpoViewController.swift:
```diff
@@ -0,0 +1,50 @@
+    lazy var expoModel: InternationalExposition = {
+        do {
+            return try JsonParser.shared.fetchData(from: .expo1900)
+        } catch {
+            fatalError("Failed to decode json data")
+        }
```
지금 프로젝트처럼 작은 프로젝트에서는 `catch` 구문이 실행될 여지가 적지만, `fatalError` 함수 호출로 에러를 처리하면 이후에 앱이 강제 종료될 여지가 있어보입니다.
간단해도 강제 종료 방식이 아닌 다른 방식으로 처리해볼 수 있을까요?😁

cms7371 replied:
JasonParser에서 fatalerror는 적절치 않은 것 같아서 에러를 정의하고 이를 이용하는 viewcontroller에서 핸들링하여 에러가 나면 에러 메세지를 표시하는 모델 또는 동작을 하도록 리팩터를 진행했습니다.
하지만 테이블뷰의 cellForRowAt 메서드에서 cell를 캐스팅을 하는 과정에서 나오는 에러는 따로 정의해 처리하는 것보다 fatalerror로 동작을 막는 것이 맞다고 생각하여 그대로 두었습니다

zdodev commented on Expo1900/Expo1900/ViewController/HeritageTableTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+        }
+    }()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        navigationItem.title = "한국의 출품작"
```
`title`을 코드로 작성하신 이유가 있을까요? 궁금합니다!🤔

cms7371 replied:
스토리보드로도 가능한것을 까먹고 있었네요☺️ 출품작 상세페이지만 코드로 하고 나머지는 수정했습니다!

zdodev commented on Expo1900/Expo1900/ViewController/HeritageTableTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+            fatalError("No matched reusable cell!")
+        }
+        let heritage = heritages[indexPath.row]
+        cell.heritageImageView.image = UIImage(named: heritage.imageName)
+        cell.heritageTitleLabel.text = heritage.name
+        cell.heritageShortDescriptionLabel.text = heritage.shortDescription
```
뷰 컨트롤러가 셀에게 데이터를 직접 설정하는 것보다, 이 책임을 셀에게 넘겨보는 것은 어떨까요?
셀이 스스로 데이터를 설정하도록 하면 역할과 책임을 더 잘 나눠볼 수 있을 것 같습니다.😁

cms7371 replied:
셀이 프로퍼티로 모델을 저장하는 것이 적절하지 않다고 생각해서 VC에서 이 책임을 맡는 것만 생각했는데,
함수 인자로 넘겨주어 설정만 하는 방법으로 깔끔하게 책임을 넘겨주는 것을 놓치고 있었네요! 좋은 제안 감사합니다!

zdodev commented on Expo1900/Expo1900/ViewController/HeritageTableTableViewController.swift:
```diff
@@ -0,0 +1,48 @@
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let detailViewController = storyboard?.instantiateViewController(withIdentifier: "heritageViewController") as? HeritageViewController else { return }
+        detailViewController.heritage = heritages[indexPath.row]
```
프로퍼티에 데이터를 넘겨보는 방식을 사용해주셨네요. 좋습니다.👍

zdodev commented on Expo1900/Expo1900/ViewController/HeritageViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        heritageDescriptionTextView.text = heritage.description
+    }
+    
+    func configureHeritageImageViewRatioConstraint(){
```
다음과 같은 오토레이아웃도 좋은 방법이라고 생각합니다!
관련 질문과 관련해서는 다음 스텝에서 오토레이아웃을 적용해보면서 같이 고민해보면 좋을 것 같습니다!

cms7371 replied:
다음 스텝에서 다시 보겠습니다!

zdodev reviewed (changes requested):
안녕하세요. @wilybear , @cms7371 !
스텝2 PR 축하드립니다~🎉

스텝2 화면 구현 조건에 따른 구현을 잘 해주셨습니다!

궁금증과 함께 몇 가지 의견을 남겨보았습니다.
궁금하신 점이 있다면 언제든지 같이 의견을 나눠보면 좋겠습니다.🤗

cms7371 committed:
refactor: Step2 PR 리뷰 반영

- Model을 Codable->Decodable로 변경
- JasonParserError 정의 및 이용하는 부분에서 핸들링 동작 추가
- JasonParser 내부 변수 네이밍 변경
- HeritageTableViewCell에 UI 변경 책임 전가
- 접근제어자 전체적으로 변경
- Navigation title 일부 스토리보드로 이동

wilybear committed:
refactor: step2 Accessibility 관련 속성 수정

- UILabel, UITextView, UIButton가 Accessibility에 맞춰
  Dynamic Font Size가 적용되도록 수정
- 메인의 PosterImageView와 HeritageViewController의 HeritageImageView
가 Accessibility에 맞춰 크기가 변경되도록 수정
- 메인 화면의 UIStackView들의 distribution 수정

zdodev commented on Expo1900/Expo1900/Utils/JsonParser.swift:
```diff
@@ -0,0 +1,44 @@
+    
+    func fetchData<T: Decodable>(from fileName: JsonFileList) throws -> T {
+        guard let assetData = NSDataAsset(name: fileName.rawValue)?.data else {
+            throw JsonParserError.noMatchedAssetName(fileName: fileName.rawValue)
```
에러 핸들링 좋습니다.👍

zdodev commented on Expo1900/Expo1900/ViewController/ExpoViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    private lazy var expo: InternationalExposition = {
+        do {
+            return try JsonParser.shared.fetchData(from: .expo1900)
+        } catch {
+            print(error.localizedDescription)
+            let errorExpo = InternationalExposition(title: "에러가 발생했습니다",
+                                                    visitors: 0,
+                                                    location: "Error",
+                                                    duration: "Error",description: error.localizedDescription)
+            return errorExpo
+        }
```
앱 강제종료가 아니라 사용자에게 에러 발생여부를 알려줄 수 있도록 변경해보셨네요. 좋습니다.👍

zdodev commented on Expo1900/Expo1900/ViewController/HeritageTableViewController.swift:
```diff
@@ -0,0 +1,52 @@
+        }
+    }()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
```
`viewDidLoad`를 특별히 override할 필요가 없다면 제거해도 좋습니다.😁

zdodev commented on Expo1900/Expo1900/ViewController/HeritageTableViewController.swift:
```diff
@@ -0,0 +1,52 @@
+            fatalError("No matched reusable cell!")
+        }
+        let heritage = heritages[indexPath.row]
+        cell.configureWithHeritage(heritage)
```
👍

zdodev reviewed (approved):
안녕하세요. @wilybear , @cms7371 !

주요 피드백과 개선된 코드를 잘 작성해주셨습니다.👍

참고가 될 만한 내용 몇 가지 적어보았습니다.
스텝2 진행하시느라 고생하셨습니다~ 이번 스텝은 여기서 마무리하겠습니다🤗
